File: JAVA/144 - Binary Tree Preorder Traversal.java
/*
 144. Binary Tree Preorder Traversal (M)

 Approach 1 : Iteration
 Time: O(n)
 Space: O(n)

 Approach 2 : Recursion
 Time: O(n)
 Space(n)

*/

/* Approach 1 */
class Solution {
    public List<Integer> preorderTraversal(TreeNode root) {
        List<Integer> res = new ArrayList<Integer>();
        Stack<TreeNode> todo = new Stack<TreeNode>();
        while (!todo.empty() || root != null) {
            while (root != null) {
                res.add(root.val);
                todo.push(root);
                root = root.left;
            }
            root = todo.pop();
            root = root.right;
        }
        return res;
    }
}

/* Approach 2 */
class Solution {
    public List<Integer> preorderTraversal(TreeNode root) {
        List<Integer> res = new ArrayList<Integer>();
        DFS(root, res);
        return res;
    }
    
    public void DFS(TreeNode root, List<Integer> res) {
        if (root == null) return;
        res.add(root.val);
        DFS(root.left, res);
        DFS(root.right, res);
    }
}

File: C++/144-Binary_Tree_Preorder_Traversal.cpp
/*
 144. Binary Tree Preorder Traversal (M)

 Recursion
 O(n),O(n)

 DFS
 O(n),O(n)

 BFS style DFS
 O(n),O(n)

*/

// Recursion
class Solution {
public:
    vector<int> preorderTraversal(TreeNode* root) {
        vector<int> res;
        DFS(root, res);
        return res;
    }
    
    void DFS(TreeNode* root, vector<int>& res) {
        if (!root) return;
        res.push_back(root->val);
        DFS(root->left, res);
        DFS(root->right, res);
    }
};

// DFS
class Solution {
public:
    vector<int> preorderTraversal(TreeNode* root) {
        vector<int> ret;
        stack<TreeNode*> todo;
        while (!todo.empty() || root) {
            while (root) {
                ret.push_back(root->val);
                todo.push(root);
                root = root->left;
            }
            root = todo.top();
            todo.pop();
            root = root->right;
        }
        return ret;
    }
};

// BFS style DFS
class Solution {
public:
    vector<int> preorderTraversal(TreeNode* root) {
        if (!root) return {};
        vector<int> res;
        stack<TreeNode*> todo;
        todo.push(root);
        while (!todo.empty()) {
            root = todo.top();
            todo.pop();
            res.push_back(root->val);
            if (root->right) todo.push(root->right);
            if (root->left) todo.push(root->left);
        }
        return res;
    }
};